lib/test/Test/Pos/Core/AddressSpec.hs

```diff
@@ -6,22 +6,91 @@ module Test.Pos.Core.AddressSpec
 
 import           Universum
 
-import           Test.Hspec            (Spec, describe)
-import           Test.Hspec.QuickCheck (modifyMaxSuccess, prop)
+import qualified Data.ByteString            as BS
+import           Formatting                 (formatToString, int, (%))
+import           Serokell.Data.Memory.Units (Byte, memory)
+import           Test.Hspec                 (Spec, describe, it, shouldBe)
+import           Test.Hspec.QuickCheck      (modifyMaxSuccess, prop)
+import           Test.QuickCheck            (Gen, arbitrary, counterexample, forAll,
+                                             frequency, vectorOf)
 
-import           Pos.Arbitrary.Core    ()
-import           Pos.Core              (IsBootstrapEraAddr (..), makePubKeyAddress,
-                                        makePubKeyHdwAddress)
-import           Pos.Crypto            (PublicKey)
-import           Pos.Crypto.HD         (HDAddressPayload (..))
+import           Pos.Arbitrary.Core         ()
+import           Pos.Binary.Class           (biSize)
+import           Pos.Core                   (Address, IsBootstrapEraAddr (..),
+                                             deriveLvl2KeyPair, largestHDAddressBoot,
+                                             largestPubKeyAddressBoot,
+                                             largestPubKeyAddressSingleKey,
+                                             makePubKeyAddress, makePubKeyAddressBoot,
+                                             makePubKeyHdwAddress)
+import           Pos.Crypto                 (EncryptedSecretKey, PassPhrase, PublicKey,
+                                             SecretKey (..), ShouldCheckPassphrase (..),
+                                             deterministicKeyGen, emptyPassphrase,
+                                             mkEncSecret, noPassEncrypt, toPublic)
+import           Pos.Crypto.HD              (HDAddressPayload (..))
 
 spec :: Spec
-spec = describe "Address" $ modifyMaxSuccess (min 10) $ do
-    prop "PK and HDW addresses with same public key are shown differently"
-         pkAndHdwAreShownDifferently
+spec = describe "Address" $ do
+    modifyMaxSuccess (min 10) $ do
+        prop "PK and HDW addresses with same public key are shown differently"
+            pkAndHdwAreShownDifferently
+
+    describe "Largest addresses" $ do
+        let genPubKeyAddrBoot = pure . makePubKeyAddressBoot . toPublic
+        largestAddressProp "PubKey address with BootstrapEra distribution"
+            genPubKeyAddrBoot largestPubKeyAddressBoot 43
+
+        let genPubKeyAddrSingleKey = pure . makePubKeyAddress
+                (IsBootstrapEraAddr False) . toPublic
+        largestAddressProp "PubKey address with SingleKey distribution"
+            genPubKeyAddrSingleKey largestPubKeyAddressSingleKey 78
+
+        let genHDAddrBoot :: SecretKey -> Gen Address
+            genHDAddrBoot sk = frequency
+                [ (1, genHDAddrBootEmptyPass sk)
```

Comment:
Wow, TIL you can do forward-bindings in a single `let` block. It's non obvious though. Maybe it makes sense to move `getHDAddrBoot` after `genHDAddress*`?